4.analyze_data/notebooks/UMAP/2.find_sc.py

```diff
@@ -0,0 +1,198 @@
+output_img_dir = pathlib.Path("./images")
+output_img_dir.mkdir(exist_ok=True)
+
+
+# ## Read in `yaml` file with settings for each single cell type
+# 
+# There is a `yaml` file called *image_settings.yaml* that holds the dictionary for all of the different UMAP single cells that we want to crop.
+# 
+# There are 4 variables per single cell:
+# - **Random state** -> number used to select the random row from the filtered UMAP data frame using the ranges set for the UMAP
+# - **UMAP0** -> range of values on the x-axis corresponding to a cluster
+# - **UMAP1** -> range of values on the y-axis corresponding to a cluster
+# - **composite_save_path** -> path to save full RGB composite image
+# - **crop_save_path** -> path to save cropped single cell RGB image
+# 
+# `UMAP0` and `UMAP1` variables are used to filter the UMAP data frame to only find single cells from specific clusters.
+
+# In[3]:
+
+
+# load in plate information
+dictionary_path = pathlib.Path(f"{settings_dir}/right_cluster.yaml")
+with open(dictionary_path) as file:
+    cell_info_dictionary = yaml.load(file, Loader=yaml.FullLoader)
+
+# Set output dirs based on the dictionary loaded in
+comp_dir = pathlib.Path(f"./images/composite_imgs/{dictionary_path.stem}")
+comp_dir.mkdir(exist_ok=True)
+crop_dir = pathlib.Path(f"./images/cropped_imgs/{dictionary_path.stem}")
+crop_dir.mkdir(exist_ok=True)
+
+# Display the first two nested dictionaries to make sure the dictionary looks correct
+pprint.pprint(list(cell_info_dictionary.items())[:2], indent=4)
+
+
+# ## Find single cells from each cell type and treatment from UMAP (plate 3)
+
+# In[4]:
+
+
+for crop_cell, info in cell_info_dictionary.items():
+    # Define your filter conditions
+    condition = (
+        (UMAP_plate3_df['Metadata_cell_type'] == crop_cell.split('-')[0]) &
+        (UMAP_plate3_df['Metadata_treatment'] == crop_cell.split('-')[1]) &
+        (UMAP_plate3_df['UMAP0'].between(info["UMAP0"][0], info["UMAP0"][1]) &
+        (UMAP_plate3_df['UMAP1'].between(info["UMAP1"][0], info["UMAP1"][1]))
+    ))
+
+    # Apply the filter
+    filtered_df = UMAP_plate3_df[condition]
+
+    # Apply the filter
+    filtered_df = UMAP_plate3_df[condition]
+
+    # Select and display only the specific columns
+    selected_columns = [
+        'Metadata_cell_type',
+        'Metadata_treatment',
+        'Metadata_Plate',
+        'Metadata_Well',
+        'Metadata_Site',
+        'Metadata_Cells_Location_Center_X',
+        'Metadata_Cells_Location_Center_Y',
+        'UMAP0',
+        'UMAP1'
+    ]
+
+    filtered_df = filtered_df[selected_columns]
+
+    # Randomly select a row using the random state parameter
+    random_row = filtered_df.sample(n=1, random_state=info["Random state"])  
+
+    # Create a filename based on Metadata_Plate, Metadata_Well, Metadata_Site
+    plate = random_row['Metadata_Plate'].values[0]
+    well = random_row['Metadata_Well'].values[0]
+    site = random_row['Metadata_Site'].values[0]
```

Comment:
Consider referencing these directly in the f-string below (instead of creating new variables for them).

Reply:
I will look into this suggestion, thank you!